src/adapter/src/catalog/apply.rs

```diff
@@ -843,15 +930,40 @@ impl CatalogState {
                     item.owner_id,
                     PrivilegeMap::from_mz_acl_items(item.privileges),
                 );
+                // If there's a matching retraction, update the new item with denormalized state
+                // from the retraction.
+                if let Some(CatalogEntry {
+                    item: _,
+                    referenced_by,
+                    used_by,
+                    id,
+                    oid: _,
+                    name: _,
+                    owner_id: _,
+                    privileges: _,
+                }) = retractions.items.remove(&key)
```

Comment:
Why is `items` the only thing that is special cased here?

Reply:
Honestly, it was mostly to avoid re-working `self.insert_item`. Ideally items would do the following when there's a matching retraction (which is what all other object types do):

  1. Update the matching retraction (`CatalogEntry`) from the durable object.
  2. Insert the retraction into `CatalogState`.

The problem is that `self.insert` does not take a `CatalogEntry`, it takes a bunch of fields that will turned into a `CatalogEntry`. I didn't want to re-work all of `self.insert` since it's used in a bunch of places. However, after thinking about it for a bit, I figured out a simple way to refactor it so it matches the above steps.